### ISSUE TITLE
Translog corruption if process is killed

### ISSUE THREAD

**Issue by masonjm** [elastic/elasticsearch#15935]
This looks like #9699, but I'm seeing it with 2.1.1.

Here's the stacktrace:

    [2016-01-12 09:57:05,517][INFO ][gateway                  ] [Googam] recovered [2] indices into cluster_state
    [2016-01-12 09:57:33,204][WARN ][indices.cluster          ] [Googam] [[audiolevel-2016.01.12][0]] marking and sending shard failed due to [failed recovery]
    [audiolevel-2016.01.12][[audiolevel-2016.01.12][0]] IndexShardRecoveryException[failed to recovery from gateway]; nested: EngineCreationFailureException[failed to recover from translog]; nested: EngineException[failed to recover from translog]; nested: ElasticsearchException[unexpected exception reading from translog snapshot of /Users/jmason/Projects/sme/elasticsearch-2.1.1/data/elasticsearch/nodes/0/indices/audiolevel-2016.01.12/0/translog/translog-33.tlog]; nested: EOFException[read past EOF. pos [107951305] length: [4] end: [107951305]];
      at org.elasticsearch.index.shard.StoreRecoveryService.recoverFromStore(StoreRecoveryService.java:254)
      at org.elasticsearch.index.shard.StoreRecoveryService.access$100(StoreRecoveryService.java:56)
      at org.elasticsearch.index.shard.StoreRecoveryService$1.run(StoreRecoveryService.java:129)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
      at java.lang.Thread.run(Thread.java:745)
    Caused by: [audiolevel-2016.01.12][[audiolevel-2016.01.12][0]] EngineCreationFailureException[failed to recover from translog]; nested: EngineException[failed to recover from translog]; nested: ElasticsearchException[unexpected exception reading from translog snapshot of /Users/jmason/Projects/sme/elasticsearch-2.1.1/data/elasticsearch/nodes/0/indices/audiolevel-2016.01.12/0/translog/translog-33.tlog]; nested: EOFException[read past EOF. pos [107951305] length: [4] end: [107951305]];
      at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:178)
      at org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:25)
      at org.elasticsearch.index.shard.IndexShard.newEngine(IndexShard.java:1408)
      at org.elasticsearch.index.shard.IndexShard.createNewEngine(IndexShard.java:1403)
      at org.elasticsearch.index.shard.IndexShard.internalPerformTranslogRecovery(IndexShard.java:906)
      at org.elasticsearch.index.shard.IndexShard.performTranslogRecovery(IndexShard.java:883)
      at org.elasticsearch.index.shard.StoreRecoveryService.recoverFromStore(StoreRecoveryService.java:245)
      ... 5 more
    Caused by: [audiolevel-2016.01.12][[audiolevel-2016.01.12][0]] EngineException[failed to recover from translog]; nested: ElasticsearchException[unexpected exception reading from translog snapshot of /Users/jmason/Projects/sme/elasticsearch-2.1.1/data/elasticsearch/nodes/0/indices/audiolevel-2016.01.12/0/translog/translog-33.tlog]; nested: EOFException[read past EOF. pos [107951305] length: [4] end: [107951305]];
      at org.elasticsearch.index.engine.InternalEngine.recoverFromTranslog(InternalEngine.java:254)
      at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:175)
      ... 11 more
    Caused by: ElasticsearchException[unexpected exception reading from translog snapshot of /Users/jmason/Projects/sme/elasticsearch-2.1.1/data/elasticsearch/nodes/0/indices/audiolevel-2016.01.12/0/translog/translog-33.tlog]; nested: EOFException[read past EOF. pos [107951305] length: [4] end: [107951305]];
      at org.elasticsearch.index.translog.TranslogReader.readSize(TranslogReader.java:102)
      at org.elasticsearch.index.translog.TranslogReader.access$000(TranslogReader.java:46)
      at org.elasticsearch.index.translog.TranslogReader$ReaderSnapshot.readOperation(TranslogReader.java:297)
      at org.elasticsearch.index.translog.TranslogReader$ReaderSnapshot.next(TranslogReader.java:290)
      at org.elasticsearch.index.translog.MultiSnapshot.next(MultiSnapshot.java:70)
      at org.elasticsearch.index.engine.InternalEngine.recoverFromTranslog(InternalEngine.java:240)
      ... 12 more
    Caused by: java.io.EOFException: read past EOF. pos [107951305] length: [4] end: [107951305]
      at org.elasticsearch.common.io.Channels.readFromFileChannelWithEofException(Channels.java:102)
      at org.elasticsearch.index.translog.ImmutableTranslogReader.readBytes(ImmutableTranslogReader.java:84)
      at org.elasticsearch.index.translog.TranslogReader.readSize(TranslogReader.java:91)
      ... 17 more

**Comment by clintongormley** [comment]
@masonjm Is this reproducible?  (It is possible that your translog was genuinely corrupted).  Also what file system are you on?

**Comment by s1monw** [comment]
@masonjm did you just upgrade you cluster or that node? It was fixed in `2.1.1` but if you hit it during shutdown before you upgrade there is not much the fix will do for you

**Comment by s1monw** [comment]
the relevant change is this https://github.com/elastic/elasticsearch/pull/15420

**Comment by s1monw** [comment]
closing - no feedbac

**Comment by theluxury** [comment]
I got it with 2.1.1. Did not just upgrade, had 2.1.1 from the start.

**Comment by s1monw** [comment]
> I just got this error with 2.1.1. Did not just upgrade, had 2.1.1 from the start.

this looks like a totally different problem. It seems like you exhausted all your filehandles? can you take a look at this https://www.elastic.co/guide/en/elasticsearch/guide/current/_file_descriptors_and_mmap.html and make sure your system is configured correctly?

**Comment by theluxury** [comment]
Yeah, from http://www.cyberciti.biz/faq/linux-increase-the-maximum-number-of-open-files/, I put `at /proc/sys/fs/file-max` and get back a huge number (1632405). Then I looked at the documents you linked and tried `sysctl -w vm.max_map_count=262144` but still get the same error.

FWIW my ES was working fine yesterday, then I installed Marvel (but couldn't get it to start since I only have ES 2.1.1), uninstalled Marvel, kill - 9 my ES, and now I have this error.

**Comment by s1monw** [comment]
when the node is running can I get the output of `curl -XGET 'http://localhost:9200/_nodes/stats/os,process?pretty=true'`

**Comment by theluxury** [comment]
```
{
  "cluster_name" : "coke-zero-crazy-delicious",
  "nodes" : {
    "rt5Be5zxSLy15WtdbAVwbg" : {
      "timestamp" : 1454788460095,
      "name" : "ip-172-31-2-14",
      "transport_address" : "ec2-52-35-132-98.us-west-2.compute.amazonaws.com/172.31.2.14:9300",
      "host" : "172.31.2.14",
      "ip" : [ "ec2-52-35-132-98.us-west-2.compute.amazonaws.com/172.31.2.14:9300", "NONE" ],
      "os" : {
        "timestamp" : 1454788460096,
        "load_average" : 1.42,
        "mem" : {
          "total_in_bytes" : 16827494400,
          "free_in_bytes" : 12977397760,
          "used_in_bytes" : 3850096640,
          "free_percent" : 77,
          "used_percent" : 23
        },
        "swap" : {
          "total_in_bytes" : 0,
          "free_in_bytes" : 0,
          "used_in_bytes" : 0
        }
      },
      "process" : {
        "timestamp" : 1454788460096,
        "open_file_descriptors" : 4091,
        "max_file_descriptors" : 4096,
        "cpu" : {
          "percent" : 71,
          "total_in_millis" : 94250
        },
        "mem" : {
          "total_virtual_in_bytes" : 23121223680
        }
      }
    }
  }
```

**Comment by s1monw** [comment]
```
        "open_file_descriptors" : 4091,
        "max_file_descriptors" : 4096,
```

I guess your settings are not applied.

**Comment by masonjm** [comment]
The original bug report was a fresh install of 2.1.1 on OSX with Java 7. It was triggered by doing a hard reset on the computer.

Since then I moved to 2.1.4 running on a VirtualBox'd Ubuntu 14.04 install with Java 8. I've gone through the same hard reset sequence several times without issue. I don't know if the problem is fixed in 2.1.4, or if it was related to that specific setup.

**Comment by theluxury** [comment]
Cool, I found out how to up that, and it's not giving me the error anymore. I'm now stuck at `[2016-02-06 20:59:01,729][INFO ][gateway                  ] [ip-172-31-2-14] recovered [98] indices into cluster_state`. I looked at my nmon and nothing seems to really be happening. Any idea how to make this faster?
